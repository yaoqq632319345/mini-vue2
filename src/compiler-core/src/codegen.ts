import { isString } from './../../shared/shared';
import {
  CREATE_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from './runtimeHelpers';
import { NodeTypes } from './ast';

export const generate = (ast) => {
  const context = createCodegenContext();
  const { push } = context;

  // 1.
  // const { toDisplayString:_toDisplayString } = Vue
  if (ast.helpers.length) {
    genFunctionPreamble(ast, context);
  }

  push('return ');

  const functionName = 'render';

  const args = ['_ctx', '_cache'];
  const signature = args.join(', ');

  push(`function ${functionName}(${signature}){ `);
  push('return ');
  genNode(ast.codegenNode, context);
  push('; }');
  return {
    code: context.code,
  };
};

// 1--- 接下来对ast进行处理，添加helpers
function genFunctionPreamble(ast: any, context) {
  const { push, helper } = context;
  const VueBinging = 'Vue';
  const aliasHelper = (s) => `${helperMapName[s]}: ${helper(s)}`;
  push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${VueBinging}`);
  push('\n');
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };

  return context;
}
function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    default:
      break;
  }
}

function genCompoundExpression(node: any, context: any) {
  const { push } = context;
  const children = node.children;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}

function genElement(node: any, context: any) {
  const { push, helper } = context;
  const { tag, children, props } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  genNodeList(genNullable([tag, props, children]), context);
  push(')');
}

function genNodeList(nodes, context) {
  const { push } = context;

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }

    if (i < nodes.length - 1) {
      push(', ');
    }
  }
}

function genNullable(args: any) {
  return args.map((arg) => arg || 'null');
}

function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(`)`);
}

function genText(node: any, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}
