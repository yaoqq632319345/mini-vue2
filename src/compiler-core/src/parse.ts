import { NodeTypes } from './ast';

type ctx = {
  source: string;
};
const enum TagType {
  Start,
  End,
}
// 1
export const baseParse = (content: string) => {
  const context = createParseContext(content);

  return createRoot(parseChildren(context, []));
};

// 4
function parseChildren(context: ctx, ancestors) {
  const nodes: any[] = [];
  // while (context.source) { // 这里不能直接这么写，因为处理子节点时，还有闭合标签</div>
  while (!isEnd(context, ancestors)) {
    let node;
    const s = context.source;
    if (s.startsWith('{{')) {
      node = parseInterpolation(context);
    } else if (s[0] === '<') {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors);
      }
    }

    if (!node) {
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
}

function isEnd(context: ctx, ancestors) {
  const s = context.source;
  if (s.startsWith('</')) {
    // 加这段更严谨的判断, 原先是遇到 </ 就结束，现在还要去栈里看看有没有开始标签
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
      }
    }
  }
  return !s;
}

function parseText(context: ctx) {
  // 2. 递归处理子节点时，需要判断结束
  let endIndex = context.source.length;
  const endTokens = ['<', '{{'];
  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);
    if (index > -1 && endIndex > index) {
      endIndex = index;
    }
  }
  const content = parseTextData(context, endIndex);
  return {
    type: NodeTypes.TEXT,
    content,
  };
}

function parseTextData(context: ctx, length: number) {
  const content = context.source.slice(0, length);
  // console.log(context.source); // some text
  advanceBy(context, length);
  // console.log(context.source); //  ''
  return content;
}

function parseElement(context: ctx, ancestors) {
  // 处理开始
  const element: any = parseTag(context, TagType.Start);
  ancestors.push(element);
  // 1. 开始标签处理完毕 递归处理子节点
  element.children = parseChildren(context, ancestors);
  ancestors.pop();
  // console.log(element.children);

  // 如果</xxx> 与 element.tag 是同一标签，则继续处理，否则抛出异常
  if (startsWithEndTagOpen(context.source, element.tag)) {
    // 处理结束
    parseTag(context, TagType.End);
  } else {
    throw new Error(`缺少结束标签:${element.tag}`);
  }
  return element;
}

function startsWithEndTagOpen(source: string, tag: string) {
  return (
    source.startsWith('</') &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}

function parseTag(context: ctx, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  // console.log(match); /* match: [ '<div', 'div', ...] */
  const tag = match[1];

  // console.log(context.source); // <div></div>
  advanceBy(context, match[0].length);
  // console.log(context.source); //  ></div>
  // TODO parse props
  advanceBy(context, 1);
  // console.log(context.source); //  </div>

  if (type === TagType.End) return;

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

// 5 处理 {{ message }}
function parseInterpolation(context: ctx) {
  const openDelimiter = '{{';
  const closeDelimiter = '}}';
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  // console.log(context.source);// {{ message }}
  advanceBy(context, openDelimiter.length);
  // console.log(context.source);//  message }}

  const rawContentLength = closeIndex - openDelimiter.length;
  // 这里优化一下，处理完之后 context.source = }}
  const rawContent = parseTextData(context, rawContentLength);
  const content = rawContent.trim();

  // console.log(context.source); // }}
  advanceBy(context, closeDelimiter.length);
  // console.log(context.source); //  ''

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

// 3
function createRoot(children: any[]) {
  return {
    children,
  };
}

// 2
function createParseContext(content: string) {
  return {
    source: content,
  };
}

// 6
function advanceBy(context: ctx, length: number) {
  context.source = context.source.slice(length);
}
