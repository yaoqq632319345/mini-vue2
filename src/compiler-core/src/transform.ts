import { TO_DISPLAY_STRING, CREATE_ELEMENT_VNODE } from './runtimeHelpers';
import { NodeTypes } from './ast';

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root);
  // 2. 给ast添加helpers
  root.helpers = [...context.helpers];
}

function createRootCodegen(root: any) {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}

function createTransformContext(root: any, options: any): any {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Set(),
    helper(key) {
      context.helpers.add(key);
    },
  };

  return context;
}
function traverseNode(node: any, context) {
  const exitFns: any = [];
  context.nodeTransforms.forEach((nodeTransform) => {
    const onExit = nodeTransform(node, context);
    // 将插件返回值作为退出执行函数
    if (onExit) exitFns.push(onExit);
  });
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ELEMENT:
    case NodeTypes.ROOT:
      traverseChildren(node, context);
      break;
    default:
      break;
  }
  // 先放入的后执行，倒着来
  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}
function traverseChildren(node: any, context: any) {
  const { children } = node;
  // 这里不用判断了
  children.forEach((child) => {
    traverseNode(child, context);
  });
}
