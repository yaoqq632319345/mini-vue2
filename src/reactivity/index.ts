export { ref, proxyRefs } from './ref';
export { effect, ReactiveEffect } from './effect';
