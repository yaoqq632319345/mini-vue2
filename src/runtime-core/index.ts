export { h } from './h';
export { renderSlots } from './helpers/renderSlots';
// 2. 解决调用render函数调用 createElementVNode 方法不存在
export { createTextVNode, createElementVNode } from './vnode';
export { getCurrentInstance, registerRuntimeCompiler } from './component';
export { provide, inject } from './apiInject';
export { createRenderer } from './renderer';
export { nextTick } from './scheduler';
// 3. 解决调用render函数调用 toDisplayString 方法不存在
export * from '../shared/toDisplayString';
