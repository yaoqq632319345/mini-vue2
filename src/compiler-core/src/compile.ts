// 2. 把parse transform generate 组装
import { generate } from './codegen';
import { baseParse } from './parse';
import { transform } from './transform';
import { transformElement } from './transforms/transformElement';
import { transformExpression } from './transforms/transformExpression';
import { transformText } from './transforms/transformText';
export function baseCompiler(template) {
  const ast = baseParse(template);
  transform(ast, {
    nodeTransforms: [
      // 1. 首先对表达式加工, {{xxx}} -> _ctx.xxx
      transformExpression,
      // 3. 最后对ele 加工, 将{ tag, props, children } 赋给 codegenNode
      transformElement,
      // 2. 其次对文本加工, hi,{{xxx}} -> 'hi,' + _ctx.xxx
      transformText,
    ],
  });
  return generate(ast);
}
