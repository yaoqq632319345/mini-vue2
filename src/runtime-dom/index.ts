import { createRenderer } from '../runtime-core/renderer';

export * from '../runtime-core';
function createElement(type) {
  return document.createElement(type);
}
function patchProp(el, k, val) {
  const isOn = (k: string) => /^on[A-Z]/.test(k);
  if (isOn(k)) {
    // 代表事件
    const event = k.slice(2).toLowerCase();
    el.addEventListener(event, val);
  } else {
    el.setAttribute(k, val);
  }
}
function insert(el, parent) {
  parent.append(el);
}
function createTextNode(text) {
  return document.createTextNode(text);
}

const renderer = createRenderer({
  createElement,
  patchProp,
  insert,
  createTextNode,
});
