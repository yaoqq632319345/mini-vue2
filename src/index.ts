export * from './runtime-dom';
export * from './reactivity';
import { baseCompiler } from './compiler-core/src';
import * as runtimeDom from './runtime-dom';

// 3. 注册编译器 compiler
import { registerRuntimeCompiler } from './runtime-dom';
function compileToFunction(template) {
  const { code } = baseCompiler(template);
  const renderFn = new Function('Vue', code)(
    /********** 
    function (Vue) {
      const { toDisplayString: _toDisplayString,  createElement: _createElement } = Vue
      return function render(_ctx, _cache) {
        return  _createElement("div", null, "hi," + _toDisplayString(_ctx.count))
      }
    }
    **********/
    runtimeDom
  );
  return renderFn;
}

registerRuntimeCompiler(compileToFunction);
