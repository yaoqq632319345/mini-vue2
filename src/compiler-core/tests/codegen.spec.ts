import { transform } from '../src/transform';
import { baseParse } from '../src/parse';
import { generate } from '../src/codegen';
// npm run test codegen -- -u
// codegen: 测试文件名
// -- 跳过npm 参数
// -u 更新快照
describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi');
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
