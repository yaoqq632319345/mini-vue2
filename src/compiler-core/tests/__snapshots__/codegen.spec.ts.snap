// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen <div><p>hi,</p>{{message}}</div> 1`] = `
"const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){ return _createElementVNode('div', null, [_createElementVNode(p, null, 'hi,'), _toDisplayString(_ctx.message)]); }"
`;

exports[`codegen <div>hi,{{message}}</div> 1`] = `
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){ return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message)); }"
`;

exports[`codegen {{message}} 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){ return _toDisplayString(_ctx.message); }"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache){ return 'hi'; }"`;
